x/sync/db.go

```diff
@@ -3,12 +3,114 @@
 
 package sync
 
-import "github.com/ava-labs/avalanchego/x/merkledb"
-
-type DB interface {
-	merkledb.Clearer
-	merkledb.MerkleRootGetter
-	merkledb.ProofGetter
-	merkledb.ChangeProofer
-	merkledb.RangeProofer
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/maybe"
+)
+
+var (
+	ErrInsufficientHistory = errors.New("insufficient history to generate proof")
+	ErrNoEndRoot           = fmt.Errorf("%w: end root not found", ErrInsufficientHistory)
+)
+
+type Marshaler[T any] interface {
+	Marshal(T) ([]byte, error)
+	Unmarshal([]byte) (T, error)
+}
+
+type DB[R any, C any] interface {
```

Comment:
nit: be consistent with this generic type name. Seems like `TRange` is used most in this PR.